### PR TITLE
Refactor replicate execution method to not iterate the origin bucket

### PR DESCRIPTION
**Pull request by yeya24**
Signed-off-by: Ben Ye <yb532204897@gmail.com>

* [x] I added CHANGELOG entry for this change.
* [ ] Change is not relevant to the end user.

## Changes

In the previous implementation, we are iterating the origin bucket, and each time, we use metaFetcher to fetch all meta files. But this seems not necessary, we can just fetch meta files once at the beginning.

### CODE REVIEW

File: pkg/replicate/scheme.go
Comment (GiedriusS) [inline pkg/replicate/scheme.go:135]: Do we really want to remove this and the following metric? May still be useful. The previous one definitely needs to go, though.

File: pkg/replicate/scheme.go
Comment (GiedriusS) [inline pkg/replicate/scheme.go:174]: Perhaps we should log the error if `metas != nil && err != nil`?

File: pkg/replicate/scheme.go
Comment (yeya24) [reply pkg/replicate/scheme.go:135]: Thanks for the quick review! This one `thanos_replicate_origin_meta_loads_total` is the same as the fetcher metric `thanos_blocks_meta_synced{state="loaded"}`

Comment (GiedriusS) [reply pkg/replicate/scheme.go:135]: Yep! You are correct. Let's document that in the `CHANGELOG.md` file and it should be good to go :+1:

File: pkg/replicate/scheme.go
Comment (yeya24) [reply pkg/replicate/scheme.go:174]: It is good to log the error here, but kind of duplicate.

I just check the code, only [here](https://github.com/thanos-io/thanos/blob/master/pkg/block/fetcher.go#L450) returns metas and error at the same time. In this case, partials != nil, so we will log the partial meta errors later.

**Comment by yeya24** [comment]
Since this pr updates replicate metrics, I also need to update our mixins. It might make sense to merge #2905 first.

**Comment by yeya24** [comment]
> AFAICT the CI failure seems related to the changes

Thanks, the previous problem was caused by the `make examples-in-container` target, and that should be fixed it in the latest commit :smile: .

File: pkg/replicate/scheme.go
Comment (OGKevin) [reply pkg/replicate/scheme.go:135]: By removing `thanos_replicate_origin_meta_loads_total`   and replacing it with `thanos_blocks_meta_synced` there is no way to distinct metrics from the 2 fetchers 🤔

Comment (OGKevin) [reply pkg/replicate/scheme.go:135]: So we either need to put back the origin metrics, or expose a label in the fetchers to distinguish them. E.g bucket name.

File: pkg/replicate/scheme.go
Comment (OGKevin) [inline pkg/replicate/scheme.go:62]: I don't think this todo should be here tho, it should be in replicator. As replicator should be able to inject labels like shipper does and not this block filter. This block filter is the last filter that will remove blocks without labels. So labels should be injected before they reach this filter.

File: pkg/replicate/scheme.go
Comment (OGKevin) [inline pkg/replicate/scheme.go:173]: Should this error not be wrapped?

File: pkg/replicate/scheme.go
Comment (yeya24) [reply pkg/replicate/scheme.go:135]: Sorry, I didn't get the point. What do you mean by `distinct metrics from the 2 fetchers`? If the fetcher metrics are exposed by 2 components, then they must have different job labels? Maybe I misunderstood your question. WDYT?

File: pkg/replicate/scheme.go
Comment (yeya24) [reply pkg/replicate/scheme.go:62]: Good catch! I agree it should be in the replicator.

File: pkg/replicate/scheme.go
Comment (yeya24) [reply pkg/replicate/scheme.go:173]: Thanks, wrapping it seems better. I will update it later.

File: pkg/replicate/scheme.go
Comment (OGKevin) [reply pkg/replicate/scheme.go:135]: > If the fetcher metrics are exposed by 2 components
this would be correct, but not in this case. In this case 2 fetchers are being used in the same component. Fetchers for the 2 buckets.

Comment (OGKevin) [reply pkg/replicate/scheme.go:135]: But maybe it is still save the expose the bucket name as a static label in the fetcher metrics?

Comment (yeya24) [reply pkg/replicate/scheme.go:135]: I don't know the answer. But this is not part of this pr and fetcher metrics are used in store and compactor as well, so if you think it is necessary to have another label we could have another issue to track it.

File: pkg/replicate/scheme.go
Comment (yeya24) [reply pkg/replicate/scheme.go:173]: Just checked it and seems it is not necessary to wrap it because the error is already wrapped in the baseFetcher. Other components (store, compact) don't wrap it as well.

**Comment by yeya24** [comment]
Can I get a review for this? I think it is close.

**Review by kakkoyun — APPROVED** [review]
lgtm

File: pkg/replicate/scheme.go
Comment (bwplotka) [reply pkg/replicate/scheme.go:135]: Yup, let's propose this in other Pr if needed

File: pkg/replicate/scheme.go
Comment (bwplotka) [reply pkg/replicate/scheme.go:173]: hm the `if err != nil && metas == nil {`  feels like code smell to me though. It just looks wrong from outside which can lead to bugs in future:

* Inside implementation can change, and metas can be produced without partial for some reason - then we will miss an error.
* Even if we say it makes sense, maybe another person will come here and refactor it anyway as it looks like error can be missed.

On other hand, shouldn't replicate FAIL on any error spotted? Even if we failed to read one block? Sometimes we heavily rely on replication status so this has to be heavily tested and controlled, ideally with failed metric being somewhere (:

File: pkg/replicate/scheme.go
Comment (bwplotka) [inline pkg/replicate/scheme.go:176]: Is this failure? This is actually valid state of blocks being potentially uploaded, no?

Comment (bwplotka) [reply pkg/replicate/scheme.go:176]: Something like `level.Info(rs.logger).Log("msg", "block meta not uploaded yet. Skipping.", "block_uuid", id.String())	
			r`

**Review by bwplotka — COMMENTED** [review]
Nice, some comments though, before merging (:

**Comment by kakkoyun** [comment]
@yeya24 This looks fine to me. You need to rebase I believe.
Ping @bwplotka are you satisfied with the state of the PR now?

File: pkg/replicate/scheme.go
Comment (bwplotka) [reply pkg/replicate/scheme.go:176]: > corrupted case, is it an error or not?

For current flow it's not error. We consider corrupted meta same as no meta.json as the reason is usually same - partial upload (:

**Review by bwplotka — APPROVED** [review]
Responded to comment, that still needs to be addressed, otherwise LGTM!

Thanks!

File: pkg/replicate/scheme.go
Comment (yeya24) [reply pkg/replicate/scheme.go]: Thanks! Updated.

File: CHANGELOG.md
Comment (bwplotka) [inline CHANGELOG.md:33]: Technically not need to mention refactor, just metric changes (:

**Review by bwplotka — APPROVED** [review]
Thanks!

Changelog entry could be better, but we can fix it later.

**Comment by OGKevin** [comment]
Whoop 🥳 🎊 Ill work on #2979.

**Comment by OGKevin** [comment]
Nice work @yeya24

**Comment by bwplotka** [comment]
:muscle:

**Pull request merged**
